### DISCUSSION TITLE
how to pass variable or object to dp.message_handler decorated function?

### DISCUSSION THREAD

**Discussion by TonyZhang03** [aiogram/aiogram#551] [Q&A] [answered]
I can't find an example provided by the package helpful to build a real chatbot. Can anyone here help to explain how or give an example?   [here](https://stackoverflow.com/questions/66884716/how-to-pass-variable-or-object-to-aiogram-dp-message-handler-decorated-function) is my problem.  Thank you in advance.

**Comment by Olegt0rr** [comment, marked as answer]
If you wanna pass a function to your handler just import it and use it directly.
If you wanna apply it for some handlers, you should use a middleware.

## Direct usage
```python
def generate_text():
    return "generated text"

async def handler(message: Message):
    text = generate_text()
    await message.answer(text)
```

## Middleware usage
#### Create a middleware
```python
from aiogram.dispatcher.middlewares import BaseMiddleware
from aiogram.types import Message

class YourMiddleware(BaseMiddleware):
    async def on_pre_process_message(self, message: Message, data: dict):
        # `text` is a name of var passed to handler
        data["text"] = generate_text()
```

#### Register it via Dispatcher (on_startup)
```python
dp.middleware.setup(YourMiddleware())
```

#### Get `text` data inside the handler
```python
from aiogram.types import Message

async def handler(message: Message, text: str):
    await message.answer(text)
```
